directly convert msg to bytevector instead of making a copy first

shengc opened the pull request:
this is to address one of the future work in wiki that aims to convert msg to bytevector directly instead of making a copy first.

shengc committed:
directly convert msg to bytevector instead of making a copy first

shengc committed:
set server.channel to be volatile since it could be read and set in multiple threads

shengc committed:
convert msg to bytevector directly in Client (instead of making a copy first)

shengc committed:
add volatile to channel var in Client

djspiewak commented on src/main/scala/scalaz/netty/Client.scala:
```diff
@@ -75,7 +76,7 @@ private[netty] final class Client(limit: Int) {
       val buf = msg.asInstanceOf[ByteBuf]
-      val bv = ByteVector(buf.nioBuffer)       // copy data (alternatives are insanely clunky)
+      val bv = ByteVector.view(buf.nioBuffer)       // copy data (alternatives are insanely clunky)
       buf.release()
```
The problem with using `view` here is that we're releasing `buf` on the very next line!  This is why I didn't write it this way in the first place.  :-)  The only way to really do a `view` here is if we handle the transcoding from `ByteVector` to the final message type (transcoding based on a `Codec` perhaps) right _here_, releasing the `ByteBuf` only after we're done.

This immediate release, btw, is the reason why the tests are failing on this branch.

shengc replied:
Ah, I did not read the code closely - yes that view operation would be busted ...

djspiewak commented on src/main/scala/scalaz/netty/Client.scala:
```diff
@@ -36,6 +36,7 @@ import _root_.io.netty.handler.codec._
 
 private[netty] final class Client(limit: Int) {
   // this isn't ugly or anything...
+  @volatile
```
Is this actually necessary?  I spent some time thinking about it, and I can't convince myself one way or another.  Remember that this is only set _once_, during initialization, after which it is purely read-only.  Given that this is a high contention reference, putting `@volatile` here probably has some appreciable performance impact.

shengc replied:
I read the code again, how about just move the channel as one of the constructor arguments for Server ? The handler class can be moved out of the scope of Server since it does not refer to the server.channel directly. The Server factory method thus would not touch the server instance until the netty channel is gotten.

shengc replied:
Ok handler refers to the server.queue. I think instead of creating queue inside the server, it should be injected into the instance instead....

djspiewak replied:
Injected into the instance?  You mean like using Guice or something?

shengc replied:
it can be managed by Guice, but what I meant is to essentially move both channel and queue to constructor. So we could move handler out of the server class scope, and channel in server does not need to be a var.

djspiewak replied:
Definitely no Guice, but I'd be interested to see a PR to this effect.  I like having `Handler` being scoped, but I'd be willing to sacrifice that to eliminate the var.

shengc replied:
This should be small amount of work - I will do it after hours. Interface of the factory method for creating Task[Server] won't be changed, so this'd be just internal.

shengc committed:
transfer bytes to Array, then to ByteVector

shengc committed:
it is safe to use ByteVector.view now

julien-truffaut commented on src/main/scala/scalaz/netty/Server.scala:
```diff
@@ -81,7 +82,14 @@ private[netty] class Server(bossGroup: NioEventLoopGroup, limit: Int) { server =
+      
+      // use a view to avoid coping the buf
+      // should be safe since there'd be no chance msg gets modified later
+//      val bv = ByteVector.view(buf.nioBuffer)       // copy data (alternatives are insanely clunky)
```
can we remove this line?

djspiewak replied:
Yes.  The comment above it should also be removed.

shengc committed:
remove comments

shengc committed:
eliminate val/var inside class (move them to constructor instead, interface of the factory method for these classes won't be changed though)

shengc commented:
I have removed those comments. I also changed interface of server/serverhandler/client/clienthandler so that we don't need to deal with the "channel" var again.

shengc commented:
thoughts, comments ? This has been pending for too long ...

djspiewak commented:
@ShengC Sorry for the delay on this.  I've been traveling all month.  I should be able to do a final review and merge this either today or tomorrow.  Thanks for all your hard work!

shengc commented:
@djspiewak absolutely no problem ... just wanna get some feedback for the commits :)
